### DISCUSSION TITLE
Vehicle Emission and Fuel consumption logs in File Output

### DISCUSSION THREAD

**Discussion by muralikrishnat29** [eclipse/mosaic#35] [Q&A] [answered]
Hello Team,

Is it possible to get the emission and fuel consumption logs of vehicles through "FileOutput" configurations? If possible, could you please let me know the config required for it? Thanks.

**Comment by kschrab** [comment, marked as answer]
It is possible to configure the FileOutput to write out fuel/emission values. For example, the `output_config.xml` of 
the Barnim scenario already includes such configuration. You would need to add the following entries to your configuration:

```xml
<subscription id="VehicleUpdates">
    <entries>
        <entry>"VEHICLE_EMISSIONS"</entry>
        <entry>Time</entry>
        <entry>Updated:Name</entry>
        <entry>Updated:VehicleEmissions.CurrentEmissions.Co2</entry>
        <entry>Updated:VehicleEmissions.CurrentEmissions.Co</entry>
        <entry>Updated:VehicleEmissions.CurrentEmissions.Hc</entry>
        <entry>Updated:VehicleEmissions.CurrentEmissions.Pmx</entry>
        <entry>Updated:VehicleEmissions.CurrentEmissions.Nox</entry>
        <entry>Updated:VehicleConsumptions.CurrentConsumptions.Fuel</entry>
        <entry>Updated:VehicleEmissions.AllEmissions.Co2</entry>
        <entry>Updated:VehicleEmissions.AllEmissions.Co</entry>
        <entry>Updated:VehicleEmissions.AllEmissions.Hc</entry>
        <entry>Updated:VehicleEmissions.AllEmissions.Pmx</entry>
        <entry>Updated:VehicleEmissions.AllEmissions.Nox</entry>
        <entry>Updated:VehicleConsumptions.AllConsumptions.Fuel</entry>
    </entries>
</subscription>
```
This is of course only an example and can be adapted to your needs.


The actual values are not calculated by MOSAIC, but by the traffic simulator SUMO. Consequently, you need to adjust the configuration of SUMO as well. Let's suppose you have various vehicle types configured in your `mapping_config.json`, each type has a name (e.g. `"Car"` in the Barnim example). For those types, the emission model needs to be configured in the SUMO configuration:

If you use the latest release 20.0, add the following line for each type you want to measure to the *.rou.xml in the `sumo` directory :

```xml
<vType id="Car" emissionClass="HBEFA3/PC_G_EU4" />
```

If you're using the current main branch instead, add the following to the `sumo_config.json`:

```json
"additionalVehicleTypeParameters": {
    "Car": {
        "emissionClass" : "HBEFA3/PC_G_EU4"
    }
}
```

The emission class should be chosen according to [this site](https://sumo.dlr.de/docs/Models/Emissions.html).

**Reply by pv2-0** [reply]
I similarly want to create an output file that is calculated by SUMO. Is there documentation to show how to specify which values I need in output_config.xml? I am interested in timeloss and waitingTime from the tripinfo output and the number of collisions from the summary output file.

**Reply by kschrab** [reply]
You can just add those output configurations to the `*.sumocfg` file like if you would do without MOSAIC. You better use absolute paths, since SUMO is executed in a temporary directory and file generated there would be removed after execution.
